### DISCUSSION TITLE
[Native DLL] Pointer to structs

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#3025] [General]
If they are in function declarations and if on C#, I will use the keyword `ref`. What should I use on PascalABC.NET? The keyword `var` or something else? Btw, what if they are in another struct? I should use `IntPtr` or `^` (real pointer)?

**Comment by SunSerega** [comment]
I already said the `var` keyword in Pascal and `ref` in C# are exactly equivalent.
But another question is if you should just blindly follow what's written in C#.
I'd say it's better to always understand why it's done the way it is.

Generally, you can translate the C pointer as an actual pointer, as a `var`-parameter, and as an array. This choice is not trivial but instead based on how this function will be used.

Pointers (or `IntPtr`) are a literal translation. But if you want to pass an address (pointer or reference) to native code, and that address points to managed memory (contents of some class) - GC might move that memory without updating the pointer, causing horrible and near-impossible to cache errors all over your program.

Arrays are the easiest to pass in/out but are not flexible.

`var`-parameters can handle diverse situations.
For instance, you can pass `a[3]` as a `var`-parameter to API expecting an array of 2 elements. That way API receives an array slice: `a[3], a[4]`.
But there are a lot of complications, like:
- If you want to pass a `NULL` address, you'll need the `nil` pointer dereference I showed earlier. But better to at least keep this black magic encapsulated somewhere close to the `external` function definition.
- You cannot always expect the `var`-parameter to be a replacement to the array:
```
## uses System.Runtime.InteropServices;

procedure p1(a: array of string); external '0.dll';
procedure p1(var a: string); external '0.dll';

var a := |'abc', 'def'|;
p1(a); // Marshals all strings in the array
p1(a[0]); // Marshals only the first element
```
`string` is non-blittable (needs marshaling), so native code doesn't receive what you pass to `p1`. It receives the output of the marshaler - the converted copy.
Same for records containing a string or a few other types, like `boolean`, `char`, and `DateTime` - these are also simple records, but they don't have a strongly defined memory layout and size in .Net, so marshaler will always try to convert them, even if conversion is a bit-wise copy.

And don't forget, unlike C - in Pascal you can always just define multiple overloads of an external function, one for each unique use case.
I usually do 1 private external method in a class and a bunch of public overloads that convert managed types the way I want, which can differ from default marshaling.
